### PR TITLE
4/n Move Accelerator into strategy - remove X_step() from accelerator

### PR DESCRIPTION
**Pull request by four4fish**
## What does this PR do?
Remove training_step() from Accelerator, and call strategy.training_step directly

In the following graph: Each color is one logic. 

[](https://postimg.cc/H8VVLFMQ)

[RFC] Should we have training_step() in Parallel Plugin, then DDPSpawning, DP and DDP doesn't need to duplicate the logic. Instead we need training_step in Horovod. 
cc: @awaelchli @ananthsub @carmocca @justusschock 

Removed TPUSpawning.training_step() as the logic is same as the super class DDPSpawn

Part of #10648

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 1]
```diff
@@ -397,8 +397,9 @@ def reduce(self, tensor, group: Optional[Any] = None, reduce_op: Union[ReduceOp,
             tensor = sync_ddp_if_available(tensor, group, reduce_op=reduce_op)
         return tensor
 
-    def training_step(self, *args, **kwargs) -> Optional[Any]:
-        return self.model(*args, **kwargs)
+    def training_step(self, step_kwargs: Dict[str, Union[Any, int]]) -> STEP_OUTPUT:
+        with self.precision_plugin.train_step_context():
+            return self.model(*step_kwargs.values())
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/ddp.py:402]: Depending on the order they were assigned to the dict, this may not be the same. Also in general, there may be keyword only arguments

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 2]
```diff
@@ -397,8 +397,9 @@ def reduce(self, tensor, group: Optional[Any] = None, reduce_op: Union[ReduceOp,
             tensor = sync_ddp_if_available(tensor, group, reduce_op=reduce_op)
         return tensor
 
-    def training_step(self, *args, **kwargs) -> Optional[Any]:
-        return self.model(*args, **kwargs)
+    def training_step(self, step_kwargs: Dict[str, Union[Any, int]]) -> STEP_OUTPUT:
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/ddp.py:400]: Can we stick to `*args` and `**kwargs` please? We can rename them to `step_args` and `step_kwargs`, but I think it's important that we have positional and keyword arguments, as this is the only way to possibly deal with [PEP 570](https://www.python.org/dev/peps/pep-0570/) and [PEP 3102](https://www.python.org/dev/peps/pep-3102/)

**Review by justusschock — CHANGES_REQUESTED** [review]

**Comment by awaelchli** [comment]
> [RFC] Should we have training_step() in Parallel Plugin, then DDPSpawning, DP and DDP doesn't need to duplicate the logic. Instead we need training_step in Horovod.

I would keep the overrides for all plugins that wrap the model with a class like DistributedDataParallel for example. This way, the definition of how the model gets wrapped lies in the plugin together with how the training_step gets called through that wrapper.

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 2 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/ddp.py:400]: @justusschock @awaelchli do you recall why the accelerator had a different signature/convention compared to the training type plugin?

**Comment by four4fish** [comment]
> > [RFC] Should we have training_step() in Parallel Plugin, then DDPSpawning, DP and DDP doesn't need to duplicate the logic. Instead we need training_step in Horovod.
> 
> I would keep the overrides for all plugins that wrap the model with a class like DistributedDataParallel for example. This way, the definition of how the model gets wrapped lies in the plugin together with how the training_step gets called through that wrapper.

I agree, even have duplicate logic is better than fail silently or inherent silencely. In the last step of the refactor, we can regrouping logics and flatten the inheritance

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 2 above)
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/ddp.py:400]: @justusschock Agree that *args, **kwargs is better, and it's align with what torch model training step will take and it's more flexible. I'm copying Accelerator's signature in this PR because all call to training_type_plugin.training_step was called through accelerator.training_step() before. And Accelerator's signature is 
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/accelerators/accelerator.py#L124-L130
If we use `*args, **kwargs` now will loose the requirement for the typing in a way, which I was concerned that could cause unexpected failures.

Comment (justusschock) [reply pytorch_lightning/plugins/training_type/ddp.py:400]: @four4fish I get it. I think we still should move on with args and kwargs :)

@ananthsub yes I think that should be part of the loops. So that plugins are unrelated to that.

Comment (four4fish) [reply pytorch_lightning/plugins/training_type/ddp.py:400]: @justusschock sounds good, will update and mark as breaking change

Comment (four4fish) [reply pytorch_lightning/plugins/training_type/ddp.py:400]: Issue created, will address the issue before update this PR
https://github.com/PyTorchLightning/pytorch-lightning/issues/10907

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/loops/optimization/optimizer_loop.py [hunk 3]
```diff
@@ -427,7 +427,7 @@ def _training_step(self, split_batch: Any, batch_idx: int, opt_idx: int) -> Clos
             )
 
             # manually capture logged metrics
-            training_step_output = self.trainer._call_accelerator_hook("training_step", *step_kwargs.values())
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimization/optimizer_loop.py:430]: After the changes in this PR, there is now only one instance of `_call_accelerator_hook` left, one that calls the `on_train_end` hook on the accelerator.
cc @daniellepintz @carmocca

**Review by awaelchli — APPROVED** [review]
nice!

File: pytorch_lightning/loops/optimization/optimizer_loop.py (hunk 3 above)
Comment (daniellepintz) [reply pytorch_lightning/loops/optimization/optimizer_loop.py:430]: do you mean `on_train_start`?
https://github.com/PyTorchLightning/pytorch-lightning/blob/3d59a2faff72aa34254b156607458f2b9acf8523/pytorch_lightning/loops/fit_loop.py#L198

Comment (awaelchli) [reply pytorch_lightning/loops/optimization/optimizer_loop.py:430]: Yes

Comment (daniellepintz) [reply pytorch_lightning/loops/optimization/optimizer_loop.py:430]: Cool. It seems like the only reason we still need that is for the `torch.cuda.empty_cache()` line here:
https://github.com/PyTorchLightning/pytorch-lightning/blob/3d59a2faff72aa34254b156607458f2b9acf8523/pytorch_lightning/accelerators/gpu.py#L50-L53

Could we put that line into `setup` instead?

**Review by carmocca — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/loops/optimization/optimizer_loop.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/loops/optimization/optimizer_loop.py:430]: I think we could do that yes.

File: pytorch_lightning/loops/optimization/manual_loop.py [hunk 4]
```diff
@@ -102,7 +102,7 @@ def advance(self, batch: Any, batch_idx: int) -> None:  # type: ignore[override]
             )
 
             # manually capture logged metrics
-            training_step_output = self.trainer._call_accelerator_hook("training_step", *step_kwargs.values())
+            training_step_output = self.trainer._call_ttp_hook("training_step", *step_kwargs.values())
```
Comment (tchaton) [inline pytorch_lightning/loops/optimization/manual_loop.py:105]: Should we use an intermediate `_training_step` which contains:

```py
def _training_step(...):
      with self.precision_plugin.train_step_context():
               self.training_step(...)
```

I am not a huge fan of duplicating the logic within all *_step methods.

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -1521,7 +1521,8 @@ def _call_ttp_hook(
         hook_name: str,
         *args: Any,
         **kwargs: Any,
-    ):
+    ) -> Optional[Any]:
```
Comment (daniellepintz) [inline pytorch_lightning/trainer/trainer.py:1524]: unfortunately this causes a mypy error, which is why I had to remove it in my PR

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -1538,7 +1538,7 @@ def _call_accelerator_hook(
         **kwargs: Any,
     ) -> Optional[Any]:
         self.lightning_module._current_fx_name = hook_name
-        fn = getattr(self.accelerator, hook_name)
+        fn = getattr(self.training_type_plugin, hook_name)
```
Comment (daniellepintz) [inline pytorch_lightning/trainer/trainer.py:1541]: why make this change? why not just call `_call_ttp_hook` instead of `_call_accelerator_hook`?

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1541]: Hey @awaelchli,

A small idea for the `test_fx_validator_integration` failure. The refactor didn't properly reset the hook name on the LightningModule as there are hooks that are nested such as the `training_step` with manual optimization. I think @carmocca has a draft PR to restore this behavior. Not sure if it could be the reason.

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1541]: This is fixed by #10575, but I don't have the time to finish it soon. If somebody wants to resolve the failing profiler tests, feel free to push to the PR.

File: pytorch_lightning/loops/optimization/manual_loop.py (hunk 4 above)
Comment (four4fish) [reply pytorch_lightning/loops/optimization/manual_loop.py]: We have a issue for simplify precision_plugins's X_steps #10891
But here, different plugins has different behaviors of call model(arg,kwarg) or model.train/test/_step(), any idea how can we simplify strategies's x_steps?

File: pytorch_lightning/trainer/trainer.py (hunk 6 above)
Comment (four4fish) [reply pytorch_lightning/trainer/trainer.py:1541]: @carmocca  you mean #10957?
